src/rust/grapl-crypto/src/aead/xchacha_blake3.rs

```diff
@@ -0,0 +1,292 @@
+use blake3::Hash;
+use chacha20::{Key, XNonce, XChaCha12};
+use chacha20::cipher::{NewCipher, StreamCipher};
+use chacha20::cipher::errors::LoopError;
+use crate::hasher::Hasher;
+
+// These domains are used to derive two separate keys from the same origin key.
+// The values can be arbitrary, but must be distinct.
+const DOMAIN_ENCRYPT: &'static [u8] = b"Grapl-12211946";
+const DOMAIN_AUTH: &'static [u8] = b"Grapl-02142019";
+
+const KEY_SIZE: usize = 32;
+const ENC_NONCE_SIZE: usize = 24;
+const MIN_KEY_NONCE_SIZE: usize = 24;
+const TOTAL_NONCE_SIZE: usize = ENC_NONCE_SIZE + MIN_KEY_NONCE_SIZE;
+
+// This aead construct is XChaCha12 with BLAKE3
+// * XChaCha12
+// * BLAKE3
+
+// Inspiration:
+// https://mccarty.io/cryptography/2021/11/29/chacha20-blake3.html
+// https://github.com/soatok/experimental-caead
+// https://soatok.blog/2020/09/09/designing-new-cryptography-for-non-standard-threat-models/
+// https://eprint.iacr.org/2019/1492.pdf
+
+#[derive(Debug, thiserror::Error)]
+pub enum AeadError {
+    #[error("AuthenticationError")]
+    AuthenticationError,
+    #[error("LoopError")]
+    LoopError
+}
+
+impl From<chacha20::cipher::errors::LoopError> for AeadError {
+    fn from(_: LoopError) -> Self {
+        Self::LoopError
+    }
+}
+
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub struct EncryptedData {
+    ciphertext: Vec<u8>,
+    pub(crate) aad: Vec<u8>,
+    #[serde(with = "crate::arrays")]
+    nonce: [u8; TOTAL_NONCE_SIZE],
+    #[serde(with = "crate::arrays")]
+    mac: [u8; KEY_SIZE],
+}
+
+/// ChaChaBlake3 is an Aead
```

Comment:
This is implied by the module structure; can something more informative be here instead?